Sources/SwiftKafka/KafkaConfig.swift

```diff
@@ -75,9 +94,48 @@ public struct KafkaConfig: Hashable, Equatable {
             }
         }
 
+        func setDeliveryReportCallback(
+            callback: @escaping (UnsafePointer<rd_kafka_message_t>?) -> Void
+        ) {
+            // Retain captured closure in this config
+            self.opaque = CapturedClosure(callback)
+
+            // Pass the captured closure to the C closure as an opaque object
+            let opaquePointer: UnsafeMutableRawPointer? = self.opaque.map { Unmanaged.passUnretained($0).toOpaque() }
+            rd_kafka_conf_set_opaque(
+                self.pointer,
+                opaquePointer
+            )
+
+            // Create a C closure that calls the captured closure
+            let callbackWrapper: (
+                @convention(c) (OpaquePointer?, UnsafePointer<rd_kafka_message_t>?, UnsafeMutableRawPointer?) -> Void
+            ) = { _, messagePointer, opaquePointer in
+
+                guard let opaquePointer = opaquePointer else {
+                    fatalError("Could not resolve reference to KafkaProducer instance")
+                }
+                let opaque = Unmanaged<CapturedClosure>.fromOpaque(opaquePointer).takeUnretainedValue()
+
+                let actualCallback = opaque.closure
+                actualCallback(messagePointer)
+            }
+
+            rd_kafka_conf_set_dr_msg_cb(
+                self.pointer,
+                callbackWrapper
+            )
```

Comment:
I think there might be a potential race if you are setting the callback a second time or nilling it out. The race is that you are setting `self.opaque` first which is going to release the current `CapturedClosure` and only at the very end you are setting the new callback in librdkafka. I think it is better if we would do the other way around and only override `self.opaque` after we set the new callback.